### PR TITLE
#2183 - Queued Assessments - Assessment Gateway Changes

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Renamed method `associateWorkflowId` to `associateWorkflowInstance` to include the part of the process to update the status to "In progress".
- Added a check to prevent the `associateWorkflowInstance` from being executed with an unexpected status (see comments in code).
- Updated assessment status to "Completed" during workflow `wrap up task`.
- Added missing audit information.
- Created a helper method to get a friendly message from the `CustomNamedError`.
- Changed the job result to cancel the workflow if the assessment is no longer in the expected status or it there is already another workflow id associated with it.

### CODE REVIEW

File: sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts [hunk 1]
```diff
@@ -49,15 +53,41 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
...
         return;
       }
       if (assessment.assessmentWorkflowId !== assessmentWorkflowId) {
         throw new CustomNamedError(
           `The assessment is already associated with another workflow instance. Current associated instance id ${assessment.assessmentWorkflowId}.`,
-          ASSESSMENT_INVALID_OPERATION_IN_THE_CURRENT_STATE,
+          ASSESSMENT_ALREADY_ASSOCIATED_WITH_DIFFERENT_WORKFLOW,
         );
       }
       // The workflow was already associated with the workflow, no need to update it again.
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts:93]: I was wondering if this part could refactored to put all the validations before the association like in: 
image

File: sources/packages/backend/libs/utilities/src/custom-named-error.ts [hunk 2]
```diff
@@ -10,4 +10,22 @@ export class CustomNamedError extends Error {
     this.objectInfo = objectInfo;
   }
   readonly objectInfo: unknown;
+
+  /**
+   * Get a user friendly error message to be displayed
+   * in the format "Message (ERROR_CODE).".
+   * @returns user friendly error message.
+   */
+  getSummaryMessage() {
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/utilities/src/custom-named-error.ts:19]: 👍

**Review by andrepestana-aot — COMMENTED** [review]
Good job! Left a question.

File: sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts:93]: I guess so, I will take a look 😉

File: sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts [hunk 3]
```diff
@@ -49,15 +53,41 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
           ASSESSMENT_NOT_FOUND,
         );
       }
+      // Check if the assessment is in one of the initial statuses. Any status different than "Submitted" and "Queued" should
+      // cancel the workflow because there is no point executing it.
+      // A workflow in these statuses can be associated and updated to "In progress" nicely. The expectation is that it would
+      // be primarily "Queued" but there is no harm in allowing "Submitted" also.
```
Comment (dheepak-aot) [inline sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts:59]: Considering submitted here is to keep the option of starting the assessments manually for any troubleshooting purpose if required right?

File: sources/packages/backend/libs/utilities/src/custom-named-error.ts (hunk 2 above)
Comment (dheepak-aot) [inline sources/packages/backend/libs/utilities/src/custom-named-error.ts:19]: return type to signature?

File: sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts:59]: Yes, you are right. The concern here was more about not allowing the association to other statuses.

File: sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts]: Method refactored, please take a look 😉

File: sources/packages/backend/libs/utilities/src/custom-named-error.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/utilities/src/custom-named-error.ts]: Added the return type.

File: sources/packages/backend/apps/workers/src/controllers/assessment/assessment.controller.ts [hunk 4]
```diff
@@ -65,29 +66,30 @@ export class AssessmentController {
...
-    const jobLogger = new Logger(job.type);
+    const logger = new Logger(job.type);
     try {
-      await this.studentAssessmentService.associateWorkflowId(
+      await this.studentAssessmentService.associateWorkflowInstance(
         job.variables.assessmentId,
         job.processInstanceKey,
       );
-      jobLogger.log("Associated the assessment id.");
+      logger.log("Associated the assessment id.");
```
Comment (ann-aot) [inline sources/packages/backend/apps/workers/src/controllers/assessment/assessment.controller.ts:75]: do we need to update the log content too?

File: sources/packages/backend/apps/workers/src/controllers/disbursement/disbursement.controller.ts [hunk 5]
```diff
@@ -102,7 +102,7 @@ export class DisbursementController {
       await this.disbursementScheduleService.associateMSFAANumber(
         job.variables.assessmentId,
       );
-      jobLogger.error("Associated the MSFAA number to the disbursements.");
+      jobLogger.log("Associated the MSFAA number to the disbursements.");
```
Comment (ann-aot) [inline sources/packages/backend/apps/workers/src/controllers/disbursement/disbursement.controller.ts:105]: 👍

File: sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts [hunk 6]
```diff
@@ -29,42 +31,74 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
...
+          throw new CustomNamedError(
+            `The assessment is already associated with another workflow instance. Current associated instance id ${assessment.assessmentWorkflowId}.`,
+            ASSESSMENT_ALREADY_ASSOCIATED_WITH_DIFFERENT_WORKFLOW,
+          );
+        }
+        // The workflow was already associated with the workflow, no need to update it again.
         throw new CustomNamedError(
-          `The assessment is already associated with another workflow instance. Current associated instance id ${assessment.assessmentWorkflowId}.`,
-          ASSESSMENT_INVALID_OPERATION_IN_THE_CURRENT_STATE,
+          `The assessment is already associated to the workflow.`,
```
Comment (ann-aot) [inline sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts:88]: double quotes?

File: sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts [hunk 7]
```diff
@@ -254,7 +288,14 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
...
   ): Promise<void> {
-    await this.repo.update({ id: assessmentId }, { workflowData });
-    // TODO: update assessment status.
+    const auditUser = await this.systemUsersService.systemUser();
+    const now = new Date();
+    await this.repo.update(assessmentId, {
+      workflowData,
+      studentAssessmentStatus: StudentAssessmentStatus.Completed,
+      studentAssessmentStatusUpdatedOn: now,
+      modifier: auditUser,
```
Comment (ann-aot) [inline sources/packages/backend/apps/workers/src/services/student-assessment/student-assessment.service.ts:297]: 👍

**Review by ann-aot — APPROVED** [review]
LGTM 👍 Good work @andrewsignori-aot . approving the PR as the added comments are very minor

**Review by dheepak-aot — APPROVED** [review]
Looks good 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

**Review by andrepestana-aot — APPROVED** [review]
Looks good! Thanks for the refactoring!

File: sources/packages/backend/apps/workers/src/controllers/assessment/assessment.controller.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/workers/src/controllers/assessment/assessment.controller.ts]: I changed as below to add more context too.
image

**Pull request merged**